### DISCUSSION TITLE
Solving obstacle nonconvex obstacle avoidance problem using SQP

### DISCUSSION THREAD

**Discussion by rtapia8484** [casadi/casadi#3886] [Q&A]
Hello all,
I would like to figure out how to solve the following nonconvex problem as an SQP, the problem is feasible and a solution is found using IPOPT.  Using sqpmethod as a solver leads to infeasibility, what settings should I adjust? This problem is using the bicycle model so ideally I think that only theta x[2] is the variable to apply the trust region? Also, how can I adjust the convergence criteria between the different SQP runs?
THank you
`
import casadi as ca
import numpy as np
from numpy import sin, cos, tan, pi
import matplotlib.pyplot as plt
import gurobipy as gp
import scipy

# General Vehicle Parameters
g = 9.8                      # Gravity Constant
m = 3600/2.2                 # Vehicle Mass
L = 2.1336                   # Wheelbase
l_fl_r = 0.85                # A/B Ratio
alfaf_max = 7*pi/180         # Design Limit
alat_max = 0.7*g             # Design Limit
K_u = 1.1                    # Understeer
b = L/(1+l_fl_r)             # CG Distance to Rear Wheel Center
a = L-b                      # CG Distance to Front Wheel Center
Iz = .5*m*a*b                # Vehicle Yaw Inertia

# Tire parameters
Cr = 5.9649e+04              # Rear tire stiffness
Cf = 6.1138e+04             # Front tire stiffness

# Constant longitudinal velocity
U = 15                       # Longitudinal Velocity [m/s]  

# Linear Bycicle model Dynamics
# [V(m/s), r(rad/s), Psi(deg)]
A = np.array([
    [-(Cf+Cr)/(m*U),(b*Cr-a*Cf)/(m*U)-U, 0],
    [(b*Cr-a*Cf)/(Iz*U),-(b**2*Cr+a**2*Cf)/(Iz*U),0],
    [0, 1, 0] ])
#B matrix with steering inputs
B = np.array([
    [Cf/m],
    [a*Cf/Iz],
    [0]])

# define casadi variables
# Fixed axis variables
v = ca.MX.sym('v')          # lateral velocity
r = ca.MX.sym('r')          # yaw rate
theta = ca.MX.sym('theta')  # angular position

th_g = ca.MX.sym('th_g')  # linearized, theta 
v_g = ca.MX.sym('v_g')    # linearized, lateral acceleration
# Global 
X = ca.MX.sym('X')          # x - global position
Y = ca.MX.sym('Y')          # y - global position

# control (steering)
u = ca.MX.sym('u')

# STATE VECTOR
x = ca.vertcat(v,r,theta,X,Y)

# define ODE 
# v_g = 0
# th_g = 0
f1 = A[0,:] @ [v,r,theta] + B[0]*u
f2 = A[1,:] @ [v,r,theta] + B[1]*u
f3 = A[2,:] @ [v,r,theta] + B[2]*u
f4 = U*cos(th_g) - v_g*sin(th_g) + (U-U)*cos(th_g) - (v-v_g)*sin(th_g) + (-U*sin(th_g)-v_g*cos(th_g))*(theta-th_g)
f5 = U*sin(th_g) + v_g*cos(th_g) + (U-U)*sin(th_g) + (v-v_g)*cos(th_g) + (U*cos(th_g)-v_g*sin(th_g))*(theta-th_g)
f6 = U*cos(theta) - v*sin(theta)
f7 = U*sin(theta) + v*cos(theta)

ode = ca.vertcat(f1,f2,f3,f4,f5)
ode2 = ca.vertcat(f1,f2,f3,f6,f7)

# Discretize system
dt = ca.MX.sym("dt")
# dt_value = T0/N
p = ca.vertcat(dt,v_g,th_g)
sys = {}
sys["x"] = x
sys["u"] = u
sys["p"] = p
sys["ode"] = ode*dt  #time scaling

intg = ca.integrator('intg','rk',sys,0,1,{"simplify":True, "number_of_finite_elements": 4})

f = ca.Function('f',[x,u,v_g,th_g],[ode],["x","u",'v_g','th_g'],["xnext"])

Fn = ca.Function('Fn',[x,u,dt,v_g,th_g],[intg(x0=x,u=u,p=ca.vertcat(dt,v_g,th_g))["xf"]],["x0","U","dt","vg","thg"],["xf"])

nx = x.numel()
nu = u.numel()

###### DEFINE CONSTRAINTS (obstacle aboidance)
# obstacle  parameters
vertexL = np.array([[-1,11,25],[-4,-2,-4]])  #[xmin;ymin]
vertexU = np.array([[1,13,27],[3.5,8,2]])   #[xmax;ymax]
xmin = vertexL[0]
xmax = vertexU[0] 
ymin = vertexL[1]
ymax = vertexU[1]
M = [52,40,22,30];                   # big-M
nobs = vertexL.shape[1]              # number of obstacles
###### OBSTACLE AVOIDANCE EQUATION ###############
rmin = np.zeros((nobs,1))
rmaj = np.zeros((nobs,1))
c_center = np.zeros((nobs,2))
for i in range(1):
    rmin[i] = (xmax[i]-xmin[i])/2
    rmaj[i] = (ymax[i]-ymin[i])/2
    c_center[i] = [(xmax[i]+xmin[i])/2,(ymax[i]+ymin[i])/2]

P1 = np.array( [ [ *(1/(rmin[0]**2)), 0],[0, *(1/(rmaj[0]**2)) ] ] )
P2 = np.array( [ [*(1/(rmin[1]**2)), 0], [0, *(1/(rmaj[1]**2))] ] )
P3 = np.array( [ [*(1/(rmin[2]**2)), 0], [0, *(1/(rmaj[2]**2))] ] )
xc = ca.MX.sym('xc')
yc = ca.MX.sym('yc')
Xbar = ca.MX.sym('Xbar')
Ybar = ca.MX.sym('Ybar')
Ps = ca.MX.sym('Ps',2,2)
eq1  = ca.vertcat(Xbar,Ybar) - ca.vertcat(xc,yc)
cons = eq1.T @ (Ps @ eq1)
# Linearize Equation
fcons = ca.Function('fcons',[ca.vertcat(Xbar,Ybar),ca.vertcat(xc,yc),Ps],[cons],['[X Y]','[xc yc]','P'],['constraint'])

################  DECLARE OPTI 
type = 'nl'
if type == 'conic':
    opti = ca.Opti('conic')
else:
    opti = ca.Opti()
#############################################  NUMBER OF NODES & TIME
N = 75
T = 3
####################################################################

####### DEFINE VARIABLES
Z = []
I = []
L = []
for k in range(N):
    Z.append(opti.variable(nx))
    I.append(opti.variable(nu))
    L.append(opti.variable())
Z.append(opti.variable(nx))
L.append(opti.variable())

Z0 = opti.parameter(nx)
T0 = opti.parameter()
XY = opti.parameter(N)

for k in range(N):

    # Multiple Shooting gap-closing constraints
    opti.subject_to( Z[k+1] == Fn(Z[k],I[k],T0/N,0,0)  )                       # IMPROVED EULER
    # opti.subject_to( Z[k+1] == Frk(Z[k],I[k],T0/N,0,0)  )                       # RK4

    # obstacle avoidance constraints
    opti.subject_to( fcons(ca.vertcat(Z[k][3],Z[k][4]),c_center[0],P1) >= 1 )

    # input constraints
    opti.subject_to( I[k] <= 30*pi/180 )
    opti.subject_to( -I[k] <= 30*pi/180 )

    # L1 norm 
    opti.subject_to( Z[k][4] <= L[k] )
    opti.subject_to( -Z[k][4] <= L[k] )

    # initialize some variables
    # opti.set_initial( Z[k], ca.vertcat(0,0,0,-40+k*5.5*N,0) )

    if k==0:
        opti.subject_to( Z[0] == Z0)

opti.subject_to( Z[N][4] <= L[N] )
opti.subject_to( -Z[N][4] <= L[N] )

Z = ca.hcat(Z)
L = ca.hcat(L)
I = ca.hcat(I)

opti.minimize(3*ca.sum2(L))

options = {}

if type =='conic':
    solver = 'gurobi'
    options = {
        'verbose': True,  # Make CasADi verbose (prints information about the solver)
        'print_time': True,  # Print time information
        # 'warm_start_init_point': 'yes',
        'gurobi': {
            'OutputFlag': 1,  # Gurobi-specific option to enable detailed logging
            'LPWarmStart': 2,  # gurobi warm start 0: none, 1: pre
            'LogFile': 'glob_SCP_ca_R2.txt',  # Save the Gurobi log to a file
            'Threads': 4,  # Example of using 4 threads (optional)
        }
    }
    opti.solver(solver,options)
elif type == 'nl':
    solver = 'ipopt'
    options = {}
    options["expand"] = True
    # options["ipopt.warm_start_init_point"] = 'yes'
    # options["ipopt.warm_start_bound_push"] = 1e-8
    # options["ipopt.warm_start_mult_bound_push"] = 1e-8
    opti.solver(solver,options)
elif type == 'qp':
    options = {}
    options["expand"] = True
    opti.solver('sqpmethod')

# initialize the state vector Z (x)
xinit = [0,0,0]
XYinit = [-15,0]
T = 2.5
# set linearization for elliptical constraint

# for i in range (3):
opti.set_value(Z0,ca.vertcat(xinit,XYinit))
# opti.set_value(XN,Xf)

opti.set_value(T0,T)
sol = opti.solve()
solver_details = sol.stats()
cpu_time = solver_details['t_wall_total']

# opti.set_initial(sol.value_variables()) #warm start (does not work with gurobi)

if True:
    try:
        Z = sol.value(Z)
    except NameError:
        Z = np.array(Z)
    xt = np.linspace(-15,35,N+1)
    r_ref = np.vstack((xt,np.zeros(N+1)))
    xobs = []
    yobs = []
    fig, ax = plt.subplots()
    ax.plot(r_ref[0,:],r_ref[1,:],'--',linewidth=0.5,label="Referece Trajectory")
    for i in range(1):
        xobs = [xmin[i],xmax[i],xmax[i],xmin[i],xmin[i]]
        yobs = [ymin[i],ymin[i],ymax[i],ymax[i],ymin[i]]
        ax.plot(xobs,yobs,linewidth = 0.75, label=f"obstacle_{i+1}")
    ax.plot(Z[3,:],Z[4,:],label = "Vehicle Path",color='blue')
    ax.legend()
`

**Comment by Nup-hsa** [comment]
Hi, i have a similar issue. I was wondering if you manage to solve the issue.

**Comment by rtapia8484** [comment]
Hello, I have not, but if I had to guess, you would manually have to formulate the problem according to SQP meaning provide the hessian etc. I instead decided to use successive convex optimization.

**Reply by brunoml5** [reply]
Hi! Can you share self contained code so that I can take a look? You can use gist ([https://gist.github.com/](https://gist.github.com/))
